### DISCUSSION TITLE
Dynamic swagger for API definition without app restart

### DISCUSSION THREAD

**Discussion by dshoem** [fastapi/fastapi#8591] [Questions] [answered]
### First Check

- [X] I added a very descriptive title to this issue.
- [X] I used the GitHub search to find a similar issue and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/samuelcolvin/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
from fastapi import FastAPI
from pydantic import BaseModel, create_model
import os
import json

app = FastAPI()

data = [(123, "John", "Smith", 16)]

def make_model():
    """
    Get the allowed rows the user accessing the API is allowed to see,
    and create a model based off of that.
    """
    script_dir = os.path.dirname(__file__)
    fp = os.path.join(script_dir, 'allowed.json')
    with open(fp, 'r') as file:
        data = json.load(file)
    

    allowed_rows = data['allowed_rows']

    fields_for_model = {}

    for field in allowed_rows:
        fields_for_model[field] = (str, None)

    model = create_model("user", **fields_for_model)
    return model


@app.get("/user", response_model=make_model())
async def fetch_user():

    # all fields present in database. If any are not in the allowed rows,
    # then they will not be added to the model created.
    fields = (
        ['id', 'first_name', 'last_name', 'building_num'])


    model = make_model()

    for i in data:
        dict_of_data = dict(zip(fields, i))

        user = model(**dict_of_data)

    return user
```

### Description

I am trying to find a way to have the API swagger schema in the UI for a user change based on what "rows" they are allowed to see. For example... user1 is allowed to see "id", "first_name", "last_name", and "building_num". So the API schema would show all of these fields in the "user" schema.

Now, lets say user2 is using the API, and they are allowed to see "id", "first_name", "last_name", but not building_num. The API schema should now exclude the building_num from the schema.

I am creating a dynamic user model for every request, based on the allowed rows found in a json file (to simulate an API call getting what the user is allowed to see for now).

Building the dynamic return model of "user" works, and when response_model is not set in the decorator, only the specified fields in the json file are shown/used to build the return model. We can remove fields from the json file without restarting the app and have the correct model built. However, when response_model is set, changing the json file does not change the "user" schema in the docs page.

Is there a way to have a dynamic API schema based on what the current user of the API is allowed to see? Any way to re-generate the schema instead of it being static and set when the API is first started?

Thanks in advance for any guidance on this

### Operating System

macOS

### FastAPI Version

0.72.0

### Python Version

3.8.9

**Comment by ddanier** [comment, marked as answer]
I don't this is the way REST APIs are supposed to work. I would recommend:
* Use different API endpoints for different user groups. Possibly providing different specs.
* Handle those permissions and set the fields users are not to see to `None` - needs those fields to be optional of course.

Everything else is the API clients / frontends job.

**Comment by EtceterisCDev** [comment]
@ddanier 

It's my first post here. 
I hope i will do it in the right way. 
I've choosen and i'm working with FastAPI for 8 monthes and i'm very very happy : the product is brilliant, the doc helps you from the beginning to an expert level.

All our APIs calls ask at minimum a specific header with a specific value, and most of them to be authenticated.
This allow us to restrict execution access to allowed users.

We want now to dynamically adapt to documentation returned ( Swagger UI or Redoc).
Depending on your profile, you can see only some routes in the documentation, the one you can use.

I can do that very easily on server variables, and i play with app.openapi and implement a custom function here.
I was able to restrict the routes, based on environments variables.

But know, depending on the header received, or on the token, i want to show a subset of routes.

I don't find a way to access the request Object in a custom_openapi function  ?

Could someone help me ?

**Comment by EtceterisCDev** [comment]
This is an extract of how we do that at the moment :

``` python
def custom_openapi():

    # settings is a global variable, defined at top level, before creating FastAPI server.
    if settings.INSTANCE_INFOS.ENVIRONMENT_NAME == 'smile':
        
        # AUthorised routes for environnement 'smile'
        PATH_POUR_SMILE = [ "/login" , 
                            "/test",
                            "/users/currentuser/graphiques",
                            "/nomenclature/rapport/{rapport_id}/paragraphe/{paragraphe_id}/image",
                            "/nomenclature/rapport/{rapport_id}/paragraphe/{paragraphe_id}/image-donnees",
                            "/nomenclature/rapport/{rapport_id}/paragraphe/{paragraphe_id}/image-html",
                            ]

        my_own_routes = []        

        # app is my FastAPI server
        for r in app.routes:
            if r.path in PATH_POUR_SMILE:
                print('--->',r.path)
                my_own_routes.append(r)

    else:
        my_own_routes = app.routes

    openapi_schema = get_openapi(
        title=settings.FASTAPI_GENERAL.PROJECT_NAME,
        version=settings.FASTAPI_GENERAL.VERSION,
        description=app.description,
        routes=my_own_routes,
    )
   
    app.openapi_schema = openapi_schema
    return app.openapi_schema
    
```

**Comment by EtceterisCDev** [comment]
Hi,

I've found a way by using this example : 

[@Kludex, Document each version on FastAPI](https://gist.github.com/Kludex/1c515aa38d22ec28d514da6b6f36da9f)
